src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -1012,6 +1012,16 @@ void AdaptDispatch::_ChangeRectAttributes(TextBuffer& textBuffer, const til::rec
                 {
                     attr.SetBackground(*changeOps.background);
                 }
+                // remove underline style if the underline is no longer set.
+                if (!attr.IsUnderlined())
+                {
+                    attr.SetUnderlineStyle(UnderlineStyle::NoUnderline);
+                }
```

Comment:
Now that the underline state can be stored in both `_attr` as `CharacterAttributes::Underlined` and in `_underlineStyle`, this kind of code worries me. If the two stop being in sync, I feel like that this could lead to annoying bugs.

Instead of code like this, would it make sense to modify `TextAttributes::SetCharacterAttributes` to update `_underlineStyle` whenever the `CharacterAttributes::Underlined` flag changes?

Reply:
Or maybe we could we remove `UnderlineStyle::NoUnderline` and only use `_attr` to decide whether underlines are active or not? That would probably work as well!

Reply:
> would it make sense to modify TextAttributes::SetCharacterAttributes to update _underlineStyle whenever the CharacterAttributes::Underlined flag changes

Yeah, I should've done this, instead of relying on a separate call to change style.

> Or maybe we could we remove `UnderlineStyle::NoUnderline` and only use `_attr` to decide whether underlines are active or not? That would probably work as well!

I believe, we only check `_attr` in `.IsUnderlined()`, didn't change anything there. The reason for keeping them in-sync is to have equality like `Attr1WithNoUnderline == Attr2WithNoUnderline` work as expected. Leaving styles random for the two attributes will result in the equality to fail.

DECRARA and DECCARA are the only functions that directly works with raw `_attr` and sets them at the end of the operation, and that is creating the out-of-sync situation.

Reply:
> Leaving styles random for the two attributes will result in the equality to fail.

For what it's worth, I probably wouldn't leave them at a random value. For instance, by setting them to `SinglyUnderlined` when `CharacterAttributes::Underlined` is reset.